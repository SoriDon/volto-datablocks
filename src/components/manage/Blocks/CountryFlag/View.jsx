import { useHistory } from 'react-router-dom';
import React from 'react';
import countryNames from './data/countries';
import './styles.less';
import { Dropdown } from 'semantic-ui-react';
import { flattenToAppURL } from '@plone/volto/helpers';

const MaybeDropdown = ({ children, countries, value, dropdown = false }) => {
  const history = useHistory();

  if (!countries || !dropdown) {
    return children;
  }

  const options = countries.map((c) => ({ text: c.title, value: c['@id'] }));

  return (
    <Dropdown
      selection
      className="countries-dd"
      text={children}
      options={options}
      defaultValue={value}
      icon="angle down"
      onChange={(_, { value }) => {
        const url = flattenToAppURL(value);
        history.push(url);
      }}
    />
  );
};

const CountryFlagView = ({ data = {}, metadata, properties }) => {
  const {
    country_name: countryCode,
    render_as,
    show_name,
    show_flag,
    show_dropdown,
  } = data;
  const Tag = render_as ? render_as.toLowerCase() : 'h2';
  const [flag, setFlag] = React.useState();
  const contentdata = metadata || properties;
  const siblings = contentdata?.['@components']?.siblings?.items || [];

  React.useEffect(() => {
    if (countryCode) {
      const code = countryCode.toLowerCase();
      import(
        /* webpackChunkName: "flags" */
        /* webpackMode: "lazy" */
        /* webpackExports: ["default", "named"] */

        `./data/svg/${code}.svg`
      ).then((module) => {
        setFlag(module.default);
      });
    }
  });
  const countryTitles = Object.values(countryNames);

  // TODO: we might as well use the Title everywhere, since we use it for the siblings
  const countries = siblings.filter((f) => countryTitles.includes(f.title));

  const countryFlag =
    (countryCode && show_flag && flag && (
      <img alt={countryNames[countryCode]} src={flag} />
    )) ||
    'no country';

  return (
    <div className="country-flag">
      {countryFlag}
      {countryCode && show_name && (
        <Tag>
          <MaybeDropdown
            dropdown={show_dropdown}
            countries={countries}
            value={countryNames[countryCode]}
          >
            {countryNames[countryCode]}
          </MaybeDropdown>
        </Tag>
      )}
    </div>
  );
};

export default CountryFlagView;
