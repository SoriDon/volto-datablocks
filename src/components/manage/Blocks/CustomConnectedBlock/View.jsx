import React from 'react';
import config from '@plone/volto/registry';
import { connectBlockToProviderData } from '../../../../hocs';
import './style.less';

const View = (props) => {
  const type = props.data.type;
  const DefaultView = () => (
    <>
      {props.mode === 'edit' && !props.data.type ? (
        <p>Please select a block type from sidebar</p>
      ) : props.mode === 'edit' && props.data.type ? (
        <p>Block type selected: {props.data.type}</p>
      ) : (
        ''
      )}
    </>
  );
  const CustomView =
    config.blocks.blocksConfig.custom_connected_block.blocks?.[type]?.view ||
    null;

  const RenderChartView = CustomView || DefaultView;

  return <RenderChartView {...props} mode={props.mode} />;
};

export default connectBlockToProviderData(View);
