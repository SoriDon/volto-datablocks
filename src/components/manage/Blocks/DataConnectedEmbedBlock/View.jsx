/**
 * View map block.
 * @module components/manage/Blocks/Maps/View
 */

import React from 'react';
import { defineMessages, injectIntl } from 'react-intl';
import cx from 'classnames';
import { connect } from 'react-redux';
import PrivacyProtection from '@eeacms/volto-embed/PrivacyProtection/PrivacyProtection';
import {
  getConnectedDataParametersForContext,
  getFilteredURL,
} from '@eeacms/volto-datablocks/helpers';
import './styles.less';

const messages = defineMessages({
  EmbededGoogleMaps: {
    id: 'Embeded Google Maps',
    defaultMessage: 'Embeded Google Maps',
  },
});

/**
 * View Embed block class.
 * @class View
 * @extends Component
 */

const ViewEmbedBlock = (props) => {
  const { data, intl } = props;

  const url = getFilteredURL(data.url, props.connected_data_parameters);

  const height =
    props?.data?.height || props.data.height === null
      ? props.data?.height
      : 600;

  return url ? (
    <PrivacyProtection data={data} height={height} {...props}>
      <p
        className={cx(
          'map-container block maps',
          {
            center: !Boolean(data.align),
          },
          data.align,
        )}
      >
        <div
          className={cx('video-inner', {
            'full-width': data.align === 'full',
          })}
          style={{
            minHeight: `${data.height || 200}px`,
          }}
        >
          <iframe
            style={{ height: `${height}px` }}
            title={intl.formatMessage(messages.EmbededGoogleMaps)}
            src={url}
            className="google-map"
            frameBorder="0"
            allowFullScreen
          />
        </div>
      </p>
    </PrivacyProtection>
  ) : (
    ''
  );
};

export default connect((state) => {
  return {
    connected_data_parameters: getConnectedDataParametersForContext(
      state?.connected_data_parameters,
      state.router?.location?.pathname,
    ),
  };
}, {})(injectIntl(ViewEmbedBlock));
