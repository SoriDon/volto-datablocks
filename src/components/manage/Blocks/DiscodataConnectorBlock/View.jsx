import React, { useState, useEffect } from 'react';
import { connect } from 'react-redux';
import { compose } from 'redux';

import { flattenToAppURL } from '@plone/volto/helpers';
import { DataConnectedValue } from '../../../../Utils';
import { SourcesBlockView } from '../../../../components';
import { setConnectedDataParameters } from '../../../../actions';
import {
  getConnectedDataParametersForProvider,
  getConnectedDataParametersForContext,
} from '../../../../helpers';

const dataParameters = (props) => {
  return (
    getConnectedDataParametersForProvider(
      props.connected_data_parameters,
      '',
    ) ||
    getConnectedDataParametersForContext(
      props.connected_data_parameters,
      props.content['@id'],
    )
  );
};

const providerView = (dataProviderKey, dataProvider, defaultDataParameters) => {
  let hasDefaultQueryParams = false;
  if (!dataProvider.queryParameterColumn || !dataProvider.queryParameterValue) {
    if (
      defaultDataParameters?.[0]?.query?.[0]?.i || //support for querystring widget, we may remove the workaround part?
      (defaultDataParameters?.[0]?.i &&
        defaultDataParameters?.[0]?.query?.[0]?.v) ||
      defaultDataParameters?.[0]?.v
    ) {
      hasDefaultQueryParams = true;
    }
  }
  return (
    <div
      key={`land-data-for-${dataProviderKey}`}
      className={dataProvider.className}
    >
      <span>
        {dataProvider.hasDiscodataConnector && (
          <DataConnectedValue
            hasQueryParammeters={
              typeof dataProvider.hasQueryParameters !== 'undefined'
                ? dataProvider.hasQueryParameters
                : true
            }
            filterIndex={hasDefaultQueryParams ? 0 : dataProviderKey}
            url={dataProvider.path}
            column={dataProvider.displayColumn}
            textTemplate={dataProvider.textTemplate}
            specifier={dataProvider.specifier}
            placeholder="_"
          />
        )}
        {dataProvider.measurmentUnit || ''}
      </span>
      {' ' + (dataProvider.additionalText || '')}
    </div>
  );
};

const bulletListView = (items) => (
  <div className="ui bulleted list">
    {items &&
      Object.entries(items).map(([key, item]) => (
        <div className="item">
          {item.leftText}
          <span className="float-right">{item.rightText}</span>
        </div>
      ))}
  </div>
);

const View = (props) => {
  const [dataProviders, setDataProviders] = useState({});
  const [parentsDataProviders, setParentsDataProviders] = useState({});
  const bulletList =
    props.data?.bullet_list?.value &&
    JSON.parse(props.data?.bullet_list?.value).properties;
  const path = flattenToAppURL(props.content['@id']);

  const updateDataProviders = () => {
    let newDataProviders = { ...dataProviders };
    if (props.data.data_providers) {
      if (
        typeof props.data.data_providers === 'object' &&
        props.data.data_providers.value
      ) {
        newDataProviders = {};
        const dataProvidersSchema =
          props.data?.data_providers?.value &&
          JSON.parse(props.data?.data_providers?.value);
        dataProvidersSchema?.fieldsets?.[0]?.fields &&
          dataProvidersSchema.fieldsets[0].fields.forEach((dataProvider) => {
            newDataProviders[dataProvider] = {
              ...dataProvidersSchema.properties[dataProvider],
            };
          });
      } else if (Array.isArray(props.data.data_providers)) {
        newDataProviders = {};
        props.data.data_providers.forEach((provider) => {
          newDataProviders[provider.id] = { ...provider };
        });
      }
    }
    setDataProviders({ ...newDataProviders });
    return newDataProviders;
  };

  const updateParentsDataProviders = () => {
    const newParentsDataProviders = {};
    dataProviders &&
      Object.entries(dataProviders).forEach(
        ([dataProviderKey, dataProvider]) => {
          if (!dataProvider.hasParent) {
            newParentsDataProviders[dataProviderKey] = { ...dataProvider };
          } else if (
            dataProvider.parent &&
            newParentsDataProviders[dataProvider.parent]
          ) {
            if (!newParentsDataProviders[dataProvider.parent].children) {
              newParentsDataProviders[dataProvider.parent].children = {};
            }
            newParentsDataProviders[dataProvider.parent].children[
              dataProviderKey
            ] = dataProvider;
          }
        },
      );
    setParentsDataProviders({ ...newParentsDataProviders });
    return newParentsDataProviders;
  };

  useEffect(() => {
    updateDataProviders();
    /* eslint-disable-next-line */
  }, []);

  useEffect(() => {
    updateParentsDataProviders();
    /* eslint-disable-next-line */
  }, [JSON.stringify(dataProviders)]);

  useEffect(() => {
    const newDataProviders = updateDataProviders();
    newDataProviders &&
      Object.entries(newDataProviders).forEach(
        ([dataProviderKey, dataProvider]) => {
          if (
            dataProvider.queryParameterColumn &&
            dataProvider.queryParameterValue
          ) {
            props.dispatch(
              setConnectedDataParameters(
                path,
                {
                  i: dataProvider.queryParameterColumn,
                  v: dataProvider.queryParameterValue,
                },
                dataProviderKey,
              ),
            );
          }
        },
      );
    /* eslint-disable-next-line */
  }, [JSON.stringify(props.data?.data_providers)]);

  const view = (
    <div className="flex h-100 pa-1">
      <div className="flex flex-column w-100">
        {/* {props.data?.block_title ? <h5>{props.data.block_title}</h5> : ''} */}
        {parentsDataProviders &&
          Object.entries(parentsDataProviders).map(
            ([dataProviderKey, dataProvider]) => {
              if (dataProvider.children) {
                return (
                  <div
                    className={dataProvider.wrapperClassName}
                    key={`data-wrapper-${dataProviderKey}`}
                  >
                    {providerView(
                      dataProviderKey,
                      dataProvider,
                      dataParameters(props),
                    )}
                    {Object.entries(
                      dataProvider.children,
                    ).map(([cildrenKey, children]) =>
                      providerView(cildrenKey, children, dataParameters(props)),
                    )}
                  </div>
                );
              }
              return (
                <div
                  className={dataProvider.wrapperClassName}
                  key={`data-wrapper-${dataProviderKey}`}
                >
                  {providerView(
                    dataProviderKey,
                    dataProvider,
                    dataParameters(props),
                  )}
                </div>
              );
            },
          )}
        {bulletList && bulletListView(bulletList)}
        {props?.data?.chartSources && (
          <div>
            <SourcesBlockView
              multipleSources={props?.data?.chartSources}
              connectorsDataProviders={dataProviders}
              download_button={props?.data?.download_button}
            />
          </div>
        )}
      </div>
    </div>
  );
  return view;
};

export default compose(
  connect((state, props) => ({
    connected_data_parameters: state.connected_data_parameters,
    content: state.content.data,
  })),
)(View);
