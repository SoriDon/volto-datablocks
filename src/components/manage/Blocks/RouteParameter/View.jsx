import React from 'react';
import { useDispatch } from 'react-redux';
import { withRouter } from 'react-router';
import { flattenToAppURL } from '@plone/volto/helpers';
import { setRouteParameter, deleteRouteParameter } from '../../../../actions';

const View = (props) => {
  const dispatch = useDispatch();
  const mounted = React.useRef(false);
  const { parameterKey = null, defaultValue = null } = props.data;
  const parameters = props.match.params;
  const contentPath = flattenToAppURL(props.properties['@id']);
  const path = props.mode === 'edit' ? contentPath : props.path;

  React.useEffect(() => {
    if (!mounted.current) return;
    if (parameterKey && defaultValue && path === contentPath) {
      dispatch(deleteRouteParameter(parameterKey));
      dispatch(
        setRouteParameter({
          i: parameterKey,
          o: 'plone.app.querystring.operation.selection.any',
          v: [parameters[parameterKey] || defaultValue],
        }),
      );
    }
    /* eslint-disable-next-line */
  }, [parameterKey, defaultValue]);

  React.useEffect(() => {
    mounted.current = true;
    if (parameterKey && defaultValue && path === contentPath) {
      dispatch(
        setRouteParameter({
          i: parameterKey,
          o: 'plone.app.querystring.operation.selection.any',
          v: [parameters[parameterKey] || defaultValue],
        }),
      );
    }

    return () => {
      mounted.current = false;
      if (parameterKey) {
        dispatch(deleteRouteParameter(parameterKey));
      }
    };
    /* eslint-disable-next-line */
  }, []);

  return (
    <>
      {props.mode === 'edit' ? (
        <>
          {!parameterKey ? <p>Set parameter key</p> : ''}
          {!defaultValue ? <p>Set default value</p> : ''}
          {parameterKey && defaultValue ? (
            <p>
              Router parameter is up and running for '{parameterKey}' with '
              {defaultValue}' as default value
            </p>
          ) : (
            ''
          )}
        </>
      ) : (
        ''
      )}
    </>
  );
};

export default withRouter(View);
