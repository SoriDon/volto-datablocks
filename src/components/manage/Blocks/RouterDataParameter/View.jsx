import React from 'react';
import { useDispatch } from 'react-redux';
import { withRouter } from 'react-router';
import { getRouterParameterValue } from './';
import {
  setRouterParameter,
  deleteRouterParameter,
} from 'volto-datablocks/actions';

const View = (props) => {
  const dispatch = useDispatch();
  const { param = null, defaultValue = null } = props.data;
  const { params = {} } = props.match;

  React.useEffect(() => {
    dispatch(
      setRouterParameter(
        param,
        getRouterParameterValue(params[param], defaultValue),
      ),
    );
    return () => {
      dispatch(deleteRouterParameter(param));
    };
    /* eslint-disable-next-line */
  }, []);

  return (
    <>
      {props.mode === 'edit' ? (
        <>
          {!param ? <p>Set parameter key</p> : ''}
          {!defaultValue ? <p>Set default value</p> : ''}
          {param && defaultValue ? (
            <p>Router data parameter is up and running for {param}</p>
          ) : (
            ''
          )}
        </>
      ) : (
        ''
      )}
    </>
  );
};

export default withRouter(View);
