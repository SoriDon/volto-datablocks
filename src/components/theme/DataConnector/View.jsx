import React, { useState } from 'react';
import { withRouter } from 'react-router';
import { useHistory } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import { Icon } from '@plone/volto/components';
import { flattenToAppURL } from '@plone/volto/helpers';
import { getRouteParameters } from 'volto-datablocks/helpers';
import { getDataFromProvider } from 'volto-datablocks/actions';
import { Container, Table, Menu } from 'semantic-ui-react';
import qs from 'querystring';

import leftSVG from '@plone/volto/icons/left-key.svg';
import rightSVG from '@plone/volto/icons/right-key.svg';

import './styles.css';

const DataConnectorView = (props) => {
  const history = useHistory();
  const dispatch = useDispatch();
  const [pagination, setPagination] = useState({
    activePage: 1,
    itemsPerPage: 20,
    totalItems: 0,
  });

  const { content = {} } = props;
  const provider_url = flattenToAppURL(props.location.pathname || '').replace(
    /\/$/,
    '',
  );

  const row_size =
    Math.min(pagination.itemsPerPage, pagination.totalItems) || 0;

  const params = useSelector((state) => {
    return (
      '?' +
      qs.stringify({
        ...qs.parse(history.location.search.replace('?', '')),
        ...getRouteParameters(
          provider_url,
          state.connected_data_parameters,
          props.match,
        ),
        p: pagination.activePage,
        nrOfHits: pagination.itemsPerPage,
      })
    );
  });

  const isPending = useSelector((state) => {
    if (provider_url === null) return false;

    const url = `${provider_url}${params}`;
    const rv = provider_url
      ? state.data_providers?.pendingConnectors?.[url]
      : false;
    return rv;
  });

  const isFailed = useSelector((state) => {
    if (provider_url === null) return false;

    const url = `${provider_url}${params}`;
    const rv = provider_url
      ? state.data_providers?.failedConnectors?.[url]
      : false;
    return rv;
  });

  const provider_data = useSelector((state) => {
    if (provider_url === null) return null;
    const url = `${provider_url}/@connector-data${params}`;
    return provider_url ? state.data_providers?.data?.[url] : null;
  });

  React.useEffect(() => {
    if (provider_url && !provider_data && !isPending && !isFailed) {
      dispatch(getDataFromProvider(provider_url, null, params));
    }
  });

  React.useEffect(() => {
    if (provider_data) {
      setPagination({
        ...pagination,
        totalItems: provider_data[Object.keys(provider_data)[0]]?.length,
      });
    }
    /* eslint-disable-next-line */
  }, [JSON.stringify(provider_data)]);

  return (
    <Container className="data-connector-view">
      <h2>{content.title}</h2>
      <pre>{content.sql_query}</pre>
      <div style={{ overflow: 'auto', width: '100%' }}>
        {provider_data && (
          <Table compact striped>
            <Table.Header>
              <Table.Row>
                {Object.keys(provider_data).map((k) => (
                  <Table.HeaderCell key={k}>{k}</Table.HeaderCell>
                ))}
              </Table.Row>
            </Table.Header>
            <Table.Body>
              {Array(row_size)
                .fill()
                .map((_, i) => (
                  <Table.Row key={i}>
                    {Object.keys(provider_data).map((k) => (
                      <Table.Cell key={`${i}-${k}`}>
                        {provider_data[k][i]}
                      </Table.Cell>
                    ))}
                  </Table.Row>
                ))}
            </Table.Body>
          </Table>
        )}
        {row_size > 0 ? (
          <Menu pagination>
            <Menu.Item
              as="a"
              icon
              disabled={pagination.activePage === 1}
              onClick={() => {
                if (pagination.activePage > 1) {
                  setPagination({
                    ...pagination,
                    activePage: pagination.activePage - 1,
                  });
                }
              }}
            >
              <Icon name={leftSVG} size="24px" />
            </Menu.Item>
            <Menu.Item
              as="a"
              icon
              disabled={row_size < pagination.itemsPerPage}
              onClick={() => {
                if (row_size === pagination.itemsPerPage) {
                  setPagination({
                    ...pagination,
                    activePage: pagination.activePage + 1,
                  });
                }
              }}
            >
              <Icon name={rightSVG} size="24px" />
            </Menu.Item>
          </Menu>
        ) : (
          ''
        )}
      </div>
    </Container>
  );
};

export default withRouter(DataConnectorView);
