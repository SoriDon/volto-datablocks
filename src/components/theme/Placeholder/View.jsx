import React from 'react';
import VisibilitySensor from 'react-visibility-sensor';

export const Placeholder = ({ children, getDOMElement, onChange, ...rest }) => {
  const [active, setActive] = React.useState(true);

  return (
    <VisibilitySensor
      scrollCheck={true}
      resizeCheck={true}
      onChange={(visible) => {
        if (visible && active) {
          setActive(false);
        }
      }}
      active={active}
      {...rest}
    >
      {({ isVisible }) => {
        if (isVisible) {
          return children();
        }

        // Without the character inside the div (non-breaking white space)
        // the size of the <div> is wrong.
        return <div className="connected-chart-placeholder">&nbsp;</div>;
      }}
    </VisibilitySensor>
  );
};

export default Placeholder;
