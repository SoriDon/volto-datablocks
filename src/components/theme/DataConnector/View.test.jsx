import React from 'react';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-intl-redux';

import View from './View';

const mockStore = configureStore();

const location = {
  pathname: '/path/to/dataconnector',
};

jest.mock('react-router', () => ({
  __esModule: true,
  withRouter: (Component) => (props) => {
    return <Component {...props} location={location} />;
  },
}));

jest.mock('react-router-dom', () => ({
  __esModule: true,
  useParams: () => {
    return {};
  },
}));

jest.mock('@plone/volto/components', () => ({
  __esModule: true,
  Icon: () => {
    return <p>This is an icon</p>;
  },
}));

test('renders DataConnector view', () => {
  const store = mockStore({
    intl: {
      locale: 'en',
      messages: {},
    },
    content: {
      data: {
        title: 'Data connector',
        sql_query: 'SELECT * FROM [FISE].[latest].[v_cnct_forest_per_capita]',
      },
    },
    connected_data_parameters: {},
    data_providers: {
      data: {
        [location.pathname]: {
          a91a94e369df59971e2e97fc57c6a4a2c661b0fb: {
            col_1: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
            col_2: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
            col_3: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
            col_4: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
          },
        },
      },
    },
  });
  const component = renderer.create(
    <Provider store={store}>
      <View location={location} content={store.content} pagination={{}} />
    </Provider>,
  );
  const json = component.toJSON();
  expect(json).toMatchSnapshot();
});
