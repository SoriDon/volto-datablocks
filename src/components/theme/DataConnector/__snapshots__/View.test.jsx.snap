// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders DataConnector view 1`] = `
<div
  className="ui container data-connector-view"
>
  <h2>
    Data connector
  </h2>
  <pre>
    SELECT * FROM [FISE].[latest].[v_cnct_forest_per_capita]
  </pre>
  <pre />
  <div
    style={
      Object {
        "overflow": "auto",
        "width": "100%",
      }
    }
  >
    <table
      className="ui striped left aligned table"
    >
      <thead
        className=""
      >
        <tr
          className=""
        >
          <th
            className=""
          >
            col_1
          </th>
          <th
            className=""
          >
            col_2
          </th>
          <th
            className=""
          >
            col_3
          </th>
          <th
            className=""
          >
            col_4
          </th>
        </tr>
      </thead>
      <tbody
        className=""
      >
        <tr
          className=""
        >
          <td
            className=""
          >
            1
          </td>
          <td
            className=""
          >
            1
          </td>
          <td
            className=""
          >
            1
          </td>
          <td
            className=""
          >
            1
          </td>
        </tr>
        <tr
          className=""
        >
          <td
            className=""
          >
            2
          </td>
          <td
            className=""
          >
            2
          </td>
          <td
            className=""
          >
            2
          </td>
          <td
            className=""
          >
            2
          </td>
        </tr>
        <tr
          className=""
        >
          <td
            className=""
          >
            3
          </td>
          <td
            className=""
          >
            3
          </td>
          <td
            className=""
          >
            3
          </td>
          <td
            className=""
          >
            3
          </td>
        </tr>
        <tr
          className=""
        >
          <td
            className=""
          >
            4
          </td>
          <td
            className=""
          >
            4
          </td>
          <td
            className=""
          >
            4
          </td>
          <td
            className=""
          >
            4
          </td>
        </tr>
        <tr
          className=""
        >
          <td
            className=""
          >
            5
          </td>
          <td
            className=""
          >
            5
          </td>
          <td
            className=""
          >
            5
          </td>
          <td
            className=""
          >
            5
          </td>
        </tr>
        <tr
          className=""
        >
          <td
            className=""
          >
            6
          </td>
          <td
            className=""
          >
            6
          </td>
          <td
            className=""
          >
            6
          </td>
          <td
            className=""
          >
            6
          </td>
        </tr>
        <tr
          className=""
        >
          <td
            className=""
          >
            7
          </td>
          <td
            className=""
          >
            7
          </td>
          <td
            className=""
          >
            7
          </td>
          <td
            className=""
          >
            7
          </td>
        </tr>
        <tr
          className=""
        >
          <td
            className=""
          >
            8
          </td>
          <td
            className=""
          >
            8
          </td>
          <td
            className=""
          >
            8
          </td>
          <td
            className=""
          >
            8
          </td>
        </tr>
        <tr
          className=""
        >
          <td
            className=""
          >
            9
          </td>
          <td
            className=""
          >
            9
          </td>
          <td
            className=""
          >
            9
          </td>
          <td
            className=""
          >
            9
          </td>
        </tr>
        <tr
          className=""
        >
          <td
            className=""
          >
            10
          </td>
          <td
            className=""
          >
            10
          </td>
          <td
            className=""
          >
            10
          </td>
          <td
            className=""
          >
            10
          </td>
        </tr>
      </tbody>
      <tfoot
        className=""
      >
        <tr
          className=""
        >
          <th
            className=""
            colSpan={4}
            style={
              Object {
                "textAlign": "center",
              }
            }
          >
            <div
              className="ui pagination menu"
            >
              <a
                className="disabled icon item"
                onClick={[Function]}
              >
                <p>
                  This is an icon
                </p>
              </a>
              <div
                className="fitted item"
                onClick={[Function]}
              >
                <div
                  className="ui tiny active disabled inline loader"
                />
              </div>
              <a
                className="icon item"
                onClick={[Function]}
              >
                <p>
                  This is an icon
                </p>
              </a>
            </div>
          </th>
        </tr>
      </tfoot>
    </table>
  </div>
</div>
`;
